contracts/finance/andromeda-mint-burn/src/execute.rs

```diff
@@ -0,0 +1,456 @@
+    if order.order_status == OrderStatus::Cancelled {
+        return Err(ContractError::CustomError {
+            msg: "Already cancelled order".to_string(),
+        });
+    }
+
+    let mut excess_amount = Uint128::zero();
+    let mut refund_nft: Option<String> = None; // To handle NFT refunds
+    let order_clone = order.clone();
+
+    for requirement in &mut order.requirements {
+        match &requirement.resource {
+            Resource::Cw20Token { cw20_addr } => {
+                if cw20_addr.clone().get_raw_address(&ctx.deps.as_ref())? == contract_addr {
+                    let user_deposit = requirement
+                        .deposits
+                        .entry(original_sender_str.clone())
+                        .or_insert(Uint128::zero());
+
+                    let remaining = requirement.amount.checked_sub(*user_deposit)?;
+
+                    if amount.le(&remaining) {
+                        *user_deposit = user_deposit.checked_add(amount)?;
+                    } else {
+                        *user_deposit = requirement.amount;
+                        excess_amount = amount.checked_sub(remaining)?;
+                    }
+                }
+            }
+            Resource::Nft {
+                cw721_addr,
+                token_id,
+            } => {
+                if cw721_addr.get_raw_address(&ctx.deps.as_ref())? == contract_addr {
+                    if let Some(received_token_id) = received_token_id.clone() {
+                        if token_id == &received_token_id {
+                            // Allow deposit of any required NFT from the same contract
+                            if !order_clone.requirements.iter().any(|r| {
+                                matches!(&r.resource, Resource::Nft { cw721_addr: addr, token_id: tid }
+                                    if addr == cw721_addr && tid == &received_token_id)
+                            }) {
+                                return Err(ContractError::CustomError {
+                                    msg: format!("Received token ID {} is not required for this order", received_token_id),
+                                });
+                            }
+
+                            let user_deposit = requirement
+                                .deposits
+                                .entry(original_sender_str.clone())
+                                .or_insert(Uint128::zero());
+
+                            let remaining = requirement.amount.checked_sub(*user_deposit)?;
+
+                            if amount.le(&remaining) {
+                                *user_deposit = user_deposit.checked_add(Uint128::one())?;
+                            } else {
+                                *user_deposit = requirement.amount;
+                                refund_nft = Some(token_id.to_string());
+                            }
+                        }
+                    }
+                }
+            }
+        }
+    }
```

Comment:
This is quite a lot of function bloat here, any way we can refactor this to individual functions?